share/availability/light/availability.go

```diff
@@ -67,77 +66,88 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
+	// load snapshot of the last sampling errors from disk
 	key := rootKey(dah)
-
 	la.dsLk.RLock()
-	exists, err := la.ds.Has(ctx, key)
+	last, err := la.ds.Get(ctx, key)
 	la.dsLk.RUnlock()
-	if err != nil || exists {
+
+	// Check for error cases
+	var samples []Sample
+	switch {
+	case err == nil && len(last) == 0:
+		// Availability has already been validated
+		return nil
+	case err != nil && !errors.Is(err, datastore.ErrNotFound):
+		// Other error occurred
 		return err
+	case errors.Is(err, datastore.ErrNotFound):
+		// No sampling result found, select new samples
+		samples, err = SampleSquare(len(dah.RowRoots), int(la.params.SampleAmount))
+		if err != nil {
+			return err
+		}
+	default:
+		// Sampling result found, unmarshal it
+		samples, err = decodeSamples(last)
+		if err != nil {
+			return err
+		}
 	}
 
-	log.Debugw("validate availability", "root", dah.String())
 	// We assume the caller of this method has already performed basic validation on the
 	// given dah/root. If for some reason this has not happened, the node should panic.
 	if err := dah.ValidateBasic(); err != nil {
 		log.Errorw("availability validation cannot be performed on a malformed DataAvailabilityHeader",
 			"err", err)
-		panic(err)
-	}
-	samples, err := SampleSquare(len(dah.RowRoots), int(la.params.SampleAmount))
-	if err != nil {
 		return err
 	}
 
 	// indicate to the share.Getter that a blockservice session should be created. This
 	// functionality is optional and must be supported by the used share.Getter.
 	ctx = getters.WithSession(ctx)
 
+	result := struct {
+		sync.Mutex
+		failed []Sample
+	}{}
+
 	log.Debugw("starting sampling session", "root", dah.String())
-	errs := make(chan error, len(samples))
+	var wg sync.WaitGroup
 	for _, s := range samples {
+		wg.Add(1)
 		go func(s Sample) {
-			log.Debugw("fetching share", "root", dah.String(), "row", s.Row, "col", s.Col)
-			_, err := la.getter.GetShare(ctx, header, s.Row, s.Col)
+			defer wg.Done()
+			// check if the sample is available
+			_, err := la.getter.GetShare(ctx, header, int(s.Row), int(s.Col))
 			if err != nil {
 				log.Debugw("error fetching share", "root", dah.String(), "row", s.Row, "col", s.Col)
-			}
-			// we don't really care about Share bodies at this point
-			// it also means we now saved the Share in local storage
-			select {
-			case errs <- err:
-			case <-ctx.Done():
+				result.Lock()
+				result.failed = append(result.failed, s)
+				result.Unlock()
 			}
 		}(s)
 	}
+	wg.Wait()
 
-	for range samples {
-		var err error
-		select {
-		case err = <-errs:
-		case <-ctx.Done():
-			err = ctx.Err()
-		}
-
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return err
-			}
-			log.Errorw("availability validation failed", "root", dah.String(), "err", err.Error())
-			if ipldFormat.IsNotFound(err) || errors.Is(err, context.DeadlineExceeded) {
-				return share.ErrNotAvailable
-			}
-			return err
-		}
+	if errors.Is(ctx.Err(), context.Canceled) {
```

Comment:
Why only context.Canceled is respected?

Reply:
I made it to align Availability interface according to: https://github.com/celestiaorg/celestia-node/blob/c93713f69b937a3d8bbba221e1c8026512e6d190/share/availability.go#L13

context.Canceled  is the only error that is not technically failed availability.